CLOUDP-149176 Update envoy-serverless to compile on CentOS 7

visualzhou opened the pull request:
This patch includes ~~4~~ 3 commits.

- Apply all-remove-tcpstats.patch to remove TCP stats for CentOS 7. This is done by running `patch -p1 <all-remove-tcpstats.patch`.
- Update certificates for tests since the existing ones expired in 2022 Aug. This is done by running `test/config/integration/certs/certs.sh`
- ~~Remove default clang paths from .bazelrc.~~
- Add evergreen scripts to compile with docker.

Only the last commit is new logic and needs thorough review. This PR is only for building Envoy and more evergreen changes will be in CLOUDP-132396.

I built the docker image using `evergreen/Dockerfile-centos7` and pushed it to `siyuanzhou/envoy-serverless-centos7` on [dockerhub](https://hub.docker.com/repository/docker/siyuanzhou/envoy-serverless-centos7).

I'm able to run the tests and build the binary with docker/podman using the following commands:
```
[ec2-user@ip-10-122-41-145 envoy-serverless]$ echo $REPODIR
/etc/envoy-serverless
[ec2-user@ip-10-122-41-145 envoy-serverless]$ docker run -it -v "$(pwd):${REPODIR}" -v "$(pwd)/envoy-build:/build" docker.io/siyuanzhou/envoy-serverless-centos7 ${REPODIR}/evergreen/7.sh
```

lydiastepanek commented on evergreen/7.sh:
```diff
@@ -0,0 +1,40 @@
+export bazel_tests="//test/common/common/... //test/integration/... //test/exe/..."
+# Include serverless specific patches.
+export bazel_tests="${bazel_tests} //test/extensions/filters/listener/proxy_protocol:proxy_protocol_test //test/extensions/common/proxy_protocol:proxy_protocol_header_test //test/extensions/transport_sockets/proxy_protocol:proxy_protocol_integration_test"
+# This test started failing in 1.17, but only when run by evergreen. not reproducible
```
does 1.17 refer to envoy version 1.17? (i know this is copied from sre comments but just curious)

visualzhou replied:
Yeah, it refers to Envoy version I believe.

lydiastepanek reviewed (commented):
LGTM, just some small questions. I'll let someone else familiar with the envoy build process be final LGTM.

JamesHeppenstall commented:
Some questions:
1. Are we adding back the TCP stats code when CentOS 7 is phased out (looking at [this comment](https://github.com/xgen-cloud/rpm-spec/blob/master/build/envoy/SPECS/envoy.spec#L45-L46))? My concern is that removing this code is going to make it difficult to upgrade to newer versions of Envoy in the future.
2. Why do we need to remove the default clang paths? Haven't we already overridden them in the CentOS 7 script? Also won't this cause issues when building on hosts that aren't CentOS 7?

I guess the crux of the above is whether or not we want to make architecture specific code changes to Envoy, rather than simply scripting around them.

milkie commented on evergreen/7.sh:
```diff
@@ -0,0 +1,40 @@
+if [[ ! -d "${BUILD_DIR}" ]]
+then
+  echo "${BUILD_DIR} mount missing - did you forget -v <something>:${BUILD_DIR}? Creating."
+  mkdir -p "${BUILD_DIR}"
```
I don't know about this.  Why not just error out?  Isn't it better to require this on the docker command line?

visualzhou replied:
This is copied from Envoy's build script. The idea is that it's fine to use the container's overlay file system to store the build artifacts.

milkie commented on evergreen/7.sh:
```diff
@@ -0,0 +1,40 @@
+  echo "${BUILD_DIR} mount missing - did you forget -v <something>:${BUILD_DIR}? Creating."
+  mkdir -p "${BUILD_DIR}"
+fi
+bazel_startup_flags="--output_base=${BUILD_DIR}"
```
Why are all the other variables exported, but not this one?

visualzhou replied:
That's a typo. Fixed. Exporting everything to make it easier to debug.

milkie commented on evergreen/7.sh:
```diff
@@ -0,0 +1,40 @@
+bazel ${bazel_startup_flags} test ${bazel_flags} ${bazel_test_flags} -- ${bazel_tests}
+bazel ${bazel_startup_flags} build ${bazel_flags} -- //:envoy
+
+# Uncomment this line to debug the docker container with -it (interactive mode)
```
Why wouldn't I just docker run "bash" and then execute 7.sh by hand?

visualzhou replied:
In my development, I'd like to have all the environment variables set up so that I can try new commands, like debugging a new specific test, but this isn't very valuable for production code. Removing this.

visualzhou commented on evergreen/7.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
Good point! Renamed.

visualzhou commented on evergreen/compile-centos7.sh:
```diff
@@ -0,0 +1,38 @@
+# in any environment other than being executed by jasper/evergreen curator.
+export bazel_tests="${bazel_tests} -//test/common/common:logger_speed_test_benchmark_test"
+export bazel_flags="--config=linux --config=clang"
+export bazel_flags="${bazel_flags} -c opt --action_env=PATH --action_env=CC=${CC} --action_env=CXX=${CXX} --action_env=LD_LIBRARY_PATH --verbose_failures -k"
```
According to Bazel's documentation, `--config=clang` at line 21 expands the option group called `clang` in place, which specifies `--action_env=CC=clang`, so a following `--action_env=CC` at line 22 should be able to override it using the value in the environment variable `CC`. However it doesn't seem to work for some external dependencies, where `CC` is correctly set to the MongoDB toolchain, but the system `clang` was called for compile. Eventually, I had to override with `--action_env=CC=${CC}`.

milkie replied:
Ugh.  Understood.

visualzhou commented:
Addressed all comments.
> Are we adding back the TCP stats code when CentOS 7 is phased out (looking at [this comment](https://github.com/xgen-cloud/rpm-spec/blob/master/build/envoy/SPECS/envoy.spec#L45-L46))? My concern is that removing this code is going to make it difficult to upgrade to newer versions of Envoy in the future.

Yes, once CentOS 7 is phased out, we will revert this patch.

> Why do we need to remove the default clang paths? Haven't we already overridden them in the CentOS 7 script? Also won't this cause issues when building on hosts that aren't CentOS 7?

Great point! I figured out how to override in the `bazel` command line. Added an inline comment to explain more.

JamesHeppenstall commented:
I'm still not convinced that simply removing TCP stats is the correct approach here. Bazel seems to have support for applying patches during the build process, which seems more idiomatic and extensible (in the event we need to apply additional patches in future builds). More so, you can add additional functionality to only apply patches for given architectures, etc. I wonder if we might want to look into this now.

Here's the link I was looking at: https://stackoverflow.com/questions/42664331/how-to-write-a-genrule-to-apply-a-patch

JamesHeppenstall commented:
> I'm still not convinced that simply removing TCP stats is the correct approach here. Bazel seems to have support for applying patches during the build process, which seems more idiomatic and extensible (in the event we need to apply additional patches in future builds). More so, you can add additional functionality to only apply patches for given architectures, etc. I wonder if we might want to look into this now.
> 
> Here's the link I was looking at: https://stackoverflow.com/questions/42664331/how-to-write-a-genrule-to-apply-a-patch

If this isn't something we want to look into now, my other request is that we split up the commit that applies the patch into a separate PR, which would make reverting it easier in the future 😄

visualzhou commented:
@JamesHeppenstall  Since we only support CentOS 7 on X86 in production, I'd like to keep the production code the same as the code in development. Thus I think it's better to keep it as a commit rather than a patch. Keeping the change as a commit will also make merging into future upstream version easier.

Totally agreed that we should keep the commits separate, that's why I had to rebase to remove the `.bazelrc` change instead of reverting it. It made code review harder, but allows us to revert the TCP stats and certificates changes easily in the future.

JamesHeppenstall commented:
> @JamesHeppenstall Since we only support CentOS 7 on X86 in production, I'd like to keep the production code the same as the code in development. Thus I think it's better to keep it as a commit rather than a patch. Keeping the change as a commit will also make merging into future upstream version easier.
> 
> Totally agreed that we should keep the commits separate, that's why I had to rebase to remove the `.bazelrc` change instead of reverting it. It made code review harder, but allows us to revert the TCP stats and certificates changes easily in the future. I'll keep all the commits and not squash them.

Sounds good to me 😄 when you merge, remember to modify the repo settings to not squash the commits together

visualzhou commented:
Signed the commits and force pushed them. @JamesHeppenstall @milkie could you please stamp?
